### PR TITLE
edge statements

### PR DESCRIPTION
**Pull request by Amber1990Zhang**

### CODE REVIEW

File: docs-2.0/3.ngql-guide/11.edge-statements/2.update-edge.md [hunk 1]
```diff
@@ -0,0 +1,24 @@
+# UPDATE EDGE
+
+```ngql
+UPDATE EDGE <src_vid> -> <dest_vid> [@rank] OF <edge_type> SET <update_columns> [WHEN <condition>] [YIELD <columns>]
+```
+
+Use `UPDATE EDGE` to update properties on an edge. The `UPDATE EDGE` statement only updates one edge type of an edge at a time.
```
Comment (randomJoe211) [inline docs-2.0/3.ngql-guide/11.edge-statements/2.update-edge.md:7]: The `UPDATE EDGE` statement only updates one edge type of an edge at a time.
=>
An `UPDATE EDGE` statement can only update the value of one property on an edge.

File: docs-2.0/3.ngql-guide/11.edge-statements/2.update-edge.md [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+
+Nebula Graph supports compare-and-set (CAS) operation.
```
Comment (randomJoe211) [inline docs-2.0/3.ngql-guide/11.edge-statements/2.update-edge.md:9]: operations, the plural form
Or just:
Nebula Graph supports compare-and-set (CAS).

File: docs-2.0/3.ngql-guide/11.edge-statements/3.upsert-edge.md [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+```ngql
+UPSERT EDGE <src_vid> -> <dest_vid> [@rank] OF <edge_type> SET <update_columns> [WHEN <condition>] [YIELD <columns>]
+```
+
+- `update_columns` is the properties of the edge to be updated. For example, `e1.col1 = $^.e1.col2 + 1` means to update `e1.col1` to `e1.col2+1`.
+
+    > **NOTE:**  `$^` indicates the edge to be updated.
+
+- `condition` is some constraints. Only when met, `UPSERT` is executed successfully. `condition` supports expression operations.
+- `columns` is the columns to be returned, `YIELD` returns the latest updated values.
```
Comment (randomJoe211) [inline docs-2.0/3.ngql-guide/11.edge-statements/3.upsert-edge.md:12]: `columns` **are/specify** the columns to be returned?

File: docs-2.0/3.ngql-guide/11.edge-statements/1.insert-edge.md [hunk 4]
```diff
@@ -0,0 +1,51 @@
+# INSERT EDGE
+
+```ngql
+INSERT EDGE <edge_type> ( <prop_name_list> ) {VALUES | VALUE}
+<src_vid> -> <dst_vid>[@<rank>] : ( <prop_value_list> )
+[, <src_vid> -> <dst_vid> : ( <prop_value_list> ), ...]
```
Comment (oldLady344) [inline docs-2.0/3.ngql-guide/11.edge-statements/1.insert-edge.md:6]: <src_vid> -> <dst_vid>[@<rank>]?

File: docs-2.0/3.ngql-guide/11.edge-statements/1.insert-edge.md [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+  [ <prop_name> [, <prop_name> ] ...]
+
+<prop_value_list> ::=
+  [ <prop_value> [, <prop_value> ] ...]
+```
+
+The `INSERT EDGE` statement inserts a (directed) edge from a source vertex (given by src_vid) to a destination vertex (given by dst_vid).
+
+* `<edge_type>` denotes the edge type, which must be created before `INSERT EDGE`.
+* `<prop_name_list>` is the property name list in the given `<edge_name>`.
```
Comment (oldLady344) [inline docs-2.0/3.ngql-guide/11.edge-statements/1.insert-edge.md:18]: in the given `<edge_type>`?

File: docs-2.0/3.ngql-guide/11.edge-statements/1.insert-edge.md [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+* `<prop_value_list>` must provide the value list according to `<prop_name_list>`. If the property values do not match the data type in the edge type, an error is returned. When the `NOT NULL` constraint is set for a given property, an error is returned if no property is given. When the default value for a property is `NULL`, you can omit to specify the property value. (TODO: create edge doc)
+* `rank` is optional, it specifies the edge rank of the same edge type. If not specified, the default value is 0.
```
Comment (oldLady344) [inline docs-2.0/3.ngql-guide/11.edge-statements/1.insert-edge.md:20]: is optional. It specifies ...

File: docs-2.0/3.ngql-guide/11.edge-statements/1.insert-edge.md [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+
+```ngql
+nebula> CREATE EDGE e2 (name string, age int);                     -- create edge type e2 with two properties
+nebula> INSERT EDGE e2 (name, age) VALUES "11"->"13":("n1", 1);          -- insert edge from "11" to "13" with two properties
+nebula> INSERT EDGE e2 (name, age) VALUES \
+"12"->"13":("n1", 1), "13"->"14":("n2", 2);                           -- insert two edges
+nebula> INSERT EDGE e2 (name, age) VALUES "11"->"13":("n1", "a13");      -- ERROR. "a13" is not int
+```
+
+An edge can be inserted/wrote multiple times. Only the last written values can be read.
```
Comment (oldLady344) [inline docs-2.0/3.ngql-guide/11.edge-statements/1.insert-edge.md:38]: inserted/written

File: docs-2.0/3.ngql-guide/11.edge-statements/2.update-edge.md (hunk 1 above)
Comment (oldLady344) [reply docs-2.0/3.ngql-guide/11.edge-statements/2.update-edge.md:7]: Only one edge type can be updated at a time with the `UPDATE EDGE` statement? 

I think we should emphasize the "only one edge type" in this sentence, so the inactive voice is OK here.

File: docs-2.0/3.ngql-guide/11.edge-statements/2.update-edge.md [hunk 8]
```diff
@@ -0,0 +1,24 @@
...
+
+> **NOTE:** `WHEN` and `YIELD` are optional.
+
+- `update_columns` is the properties of the edge to be updated. For example, `e1.col1 = $^.e1.col2 + 1` means to update `e1.col1` to `e1.col2+1`.
+
+    > **NOTE:**  `$^` indicates the vertex to be updated.
```
Comment (oldLady344) [inline docs-2.0/3.ngql-guide/11.edge-statements/2.update-edge.md:15]: vertex?

File: docs-2.0/3.ngql-guide/11.edge-statements/2.update-edge.md [hunk 9]
```diff
@@ -0,0 +1,24 @@
...
+
+- `condition` is some constraints, only when met, `UPDATE` is executed successfully. `condition` supports expression operations.
```
Comment (oldLady344) [inline docs-2.0/3.ngql-guide/11.edge-statements/2.update-edge.md:17]: . Only when the `condition` is met, `UPDATE` ...

File: docs-2.0/3.ngql-guide/11.edge-statements/2.update-edge.md [hunk 10]
```diff
@@ -0,0 +1,24 @@
...
+- `columns` is the columns to be returned, `YIELD` returns the latest updated values.
```
Comment (oldLady344) [inline docs-2.0/3.ngql-guide/11.edge-statements/2.update-edge.md:18]: `YIELD <columns>` returns the properties with updated values?

File: docs-2.0/3.ngql-guide/11.edge-statements/2.update-edge.md [hunk 11]
```diff
@@ -0,0 +1,24 @@
+# UPDATE EDGE
+
+```ngql
+UPDATE EDGE <src_vid> -> <dest_vid> [@rank] OF <edge_type> SET <update_columns> [WHEN <condition>] [YIELD <columns>]
```
Comment (oldLady344) [inline docs-2.0/3.ngql-guide/11.edge-statements/2.update-edge.md:4]: can we replace "columns" with "properties" in this statment? 

In Nebula Graph, "property" should be emphasized?

File: docs-2.0/3.ngql-guide/11.edge-statements/3.upsert-edge.md [hunk 12]
```diff
@@ -0,0 +1,42 @@
+# UPSERT EDGE
+
+```ngql
+UPSERT EDGE <src_vid> -> <dest_vid> [@rank] OF <edge_type> SET <update_columns> [WHEN <condition>] [YIELD <columns>]
```
Comment (oldLady344) [inline docs-2.0/3.ngql-guide/11.edge-statements/3.upsert-edge.md:4]: same question: can we replace "columns" with "properties" or "prop" in this statement?

File: docs-2.0/3.ngql-guide/11.edge-statements/1.insert-edge.md [hunk 13]
```diff
@@ -0,0 +1,51 @@
...
+-- insert edge with the new values.
+nebula> INSERT EDGE e2 (name, age) VALUES "11"->"13":("n1", 12);
+nebula> INSERT EDGE e2 (name, age) VALUES "11"->"13":("n1", 13);
+nebula> INSERT EDGE e2 (name, age) VALUES "11"->"13":("n1", 14);
+nebula> FETCH PROP ON e2 "11"->"13";
++---------+---------+----------+---------+--------+
+| e2._src | e2._dst | e2._rank | e2.name | e2.age |
++---------+---------+----------+---------+--------+
+| "11"    | "13"    | 0        | "n1"    | 14     |
++---------+---------+----------+---------+--------+      -- Only the last version can be read
```
Comment (oldLady344) [inline docs-2.0/3.ngql-guide/11.edge-statements/1.insert-edge.md:50]: Only the last write can be read?

File: docs-2.0/3.ngql-guide/11.edge-statements/3.upsert-edge.md [hunk 14]
```diff
@@ -0,0 +1,42 @@
...
+```
+
+- `update_columns` is the properties of the edge to be updated. For example, `e1.col1 = $^.e1.col2 + 1` means to update `e1.col1` to `e1.col2+1`.
+
+    > **NOTE:**  `$^` indicates the edge to be updated.
+
+- `condition` is some constraints. Only when met, `UPSERT` is executed successfully. `condition` supports expression operations.
```
Comment (oldLady344) [inline docs-2.0/3.ngql-guide/11.edge-statements/3.upsert-edge.md:11]: Only when the constraints are met, ...

File: docs-2.0/3.ngql-guide/11.edge-statements/3.upsert-edge.md [hunk 15]
```diff
@@ -0,0 +1,42 @@
...
+- `columns` is the columns to be returned, `YIELD` returns the latest updated values.
+
+`UPSERT` is a combination of `UPDATE` and `INSERT`.  Use `UPSERT EDGE` to update properties on an edge if it exists or insert a new edge if it does not exist. The `UPDATE EDGE` statement only updates one edge type of an edge at a time.
```
Comment (oldLady344) [inline docs-2.0/3.ngql-guide/11.edge-statements/3.upsert-edge.md:14]: .  Use [ extra space]

File: docs-2.0/3.ngql-guide/11.edge-statements/1.insert-edge.md [hunk 16]
```diff
@@ -0,0 +1,51 @@
...
+<prop_name_list> ::=
+  [ <prop_name> [, <prop_name> ] ...]
+
+<prop_value_list> ::=
+  [ <prop_value> [, <prop_value> ] ...]
+```
+
+The `INSERT EDGE` statement inserts a (directed) edge from a source vertex (given by src_vid) to a destination vertex (given by dst_vid).
+
+* `<edge_type>` denotes the edge type, which must be created before `INSERT EDGE`.
```
Comment (oldLady344) [inline docs-2.0/3.ngql-guide/11.edge-statements/1.insert-edge.md:17]: Should we mention that only one edge type can be specified in this statement?

File: docs-2.0/3.ngql-guide/11.edge-statements/3.upsert-edge.md (hunk 15 above)
Comment (oldLady344) [reply docs-2.0/3.ngql-guide/11.edge-statements/3.upsert-edge.md:14]: Only one edge type can be updated or inserted at a time with the `UPSERT EDGE` statement.

File: docs-2.0/3.ngql-guide/11.edge-statements/3.upsert-edge.md [hunk 17]
```diff
@@ -0,0 +1,42 @@
...
+
+The performance of `UPSERT` is much lower than that of `INSERT`, because `UPSERT` is a read-modify-write serialization operation at the partition level.
+
+>**DON'T:** DO NOT use `UPSERT` for scenarios with highly concurrent writes.
+
+- If the edge does not exist, a new edge is created no matter whether the condition in the `WHEN` clause is met or not. The property columns not specified by the `SET` statement use the default values of the columns. If there are no default values, an error is returned.
```
Comment (oldLady344) [inline docs-2.0/3.ngql-guide/11.edge-statements/3.upsert-edge.md:20]: , a new edge is created no matter whether the condition in the `WHEN` clause is met or not. ==> does that mean "the condition in the `WHEN` clause is ignored and a new edge is created"?

Comment (oldLady344) [inline docs-2.0/3.ngql-guide/11.edge-statements/3.upsert-edge.md:20]: The properties not specified by the `SET` statement use their default values?

File: docs-2.0/3.ngql-guide/11.edge-statements/3.upsert-edge.md [hunk 18]
```diff
@@ -0,0 +1,42 @@
...
+- If the edge exists and the `WHEN` condition is not met, Nebula Graph does nothing.
+
+Consider the following example:
+
+```ngql
+nebula> INSERT EDGE serve(start_year, end_year) VALUES "player100" -> "team200":(1997, 2016);  -- Insert a new edge.
+
+nebula> UPSERT EDGE "player100" -> "team200" OF serve SET start_year = serve.start_year + 2 WHEN serve.end_year == 2016 YIELD serve.start_year AS Start, serve.end_year AS End;
++-------+------+
+| Start | end  |
```
Comment (oldLady344) [inline docs-2.0/3.ngql-guide/11.edge-statements/3.upsert-edge.md:31]: end OR End?

File: docs-2.0/3.ngql-guide/11.edge-statements/4.delete-edge.md [hunk 19]
```diff
@@ -0,0 +1,24 @@
+# DELETE EDGE
+
+```ngql
+DELETE EDGE <edge_type> <vid> -> <vid>[@<rank>]
```
Comment (oldLady344) [inline docs-2.0/3.ngql-guide/11.edge-statements/4.delete-edge.md:4]: <src_vid> -> <dst_vid>?

File: docs-2.0/3.ngql-guide/11.edge-statements/3.upsert-edge.md (hunk 12 above)
Comment (oldLady344) [inline docs-2.0/3.ngql-guide/11.edge-statements/3.upsert-edge.md:4]: <dest_vid> --> <dst_vid>

File: docs-2.0/3.ngql-guide/11.edge-statements/4.delete-edge.md [hunk 20]
```diff
@@ -0,0 +1,24 @@
...
+```
+
+Use `DELETE EDGE` to delete edges. The `DELETE EDGE` statement deletes one vertex or multiple edges at a time. You can use `DELETE EDGE` together with pipe. For more information on pipe, see [Pipe operator](../5.operators/4.pipe.md).
```
Comment (oldLady344) [inline docs-2.0/3.ngql-guide/11.edge-statements/4.delete-edge.md:7]: vertex?

Comment (oldLady344) [inline docs-2.0/3.ngql-guide/11.edge-statements/4.delete-edge.md:7]: For more information about pipe,

File: docs-2.0/3.ngql-guide/11.edge-statements/4.delete-edge.md [hunk 21]
```diff
@@ -0,0 +1,24 @@
...
+
+## Examples
+
+```ngql
+nebula> DELETE EDGE serve "player100" -> "team200";
+```
+
+This query deletes the edge whose edge type is serve, source vertex is "player100", destination vertex is "team200".
```
Comment (oldLady344) [inline docs-2.0/3.ngql-guide/11.edge-statements/4.delete-edge.md:15]: the edge whose edge type is serve, source vertex is "player100", destination vertex is "team200". --> the `serve` edge from "player100" to "team200"?

File: docs-2.0/3.ngql-guide/11.edge-statements/4.delete-edge.md (hunk 19 above)
Comment (oldLady344) [inline docs-2.0/3.ngql-guide/11.edge-statements/4.delete-edge.md:4]: if more than one edges can be deleted, should we update this syntax with "DELETE EDGE <edge_type> <src_vid> -> <dst_vid> [, <src_vid> -> <dst_vid> ...]"? And, can "rank" be deleted?

File: docs-2.0/3.ngql-guide/11.edge-statements/2.update-edge.md (hunk 1 above)
Comment (Amber1990Zhang) [reply docs-2.0/3.ngql-guide/11.edge-statements/2.update-edge.md:7]: The `UPDATE EDGE` statement only updates one edge at a time.
I prefer this

Comment (oldLady344) [reply docs-2.0/3.ngql-guide/11.edge-statements/2.update-edge.md:7]: > The `UPDATE EDGE` statement only updates one edge at a time.
> I prefer this

one edge type, not one edge. And, in your version, I think "updates only one edge" is clearer. "only" should be next to the word that it modifies.

File: docs-2.0/3.ngql-guide/11.edge-statements/3.upsert-edge.md (hunk 17 above)
Comment (Amber1990Zhang) [reply docs-2.0/3.ngql-guide/11.edge-statements/3.upsert-edge.md:20]: You can take it this way.

**Review by randomJoe211 — APPROVED** [review]

File: docs-2.0/3.ngql-guide/11.edge-statements/1.insert-edge.md [hunk 22]
```diff
@@ -0,0 +1,51 @@
...
+[, <src_vid> -> <dst_vid>[@<rank>] : ( <prop_value_list> ), ...]
+
+<prop_name_list> ::=
+  [ <prop_name> [, <prop_name> ] ...]
+
+<prop_value_list> ::=
+  [ <prop_value> [, <prop_value> ] ...]
+```
+
+The `INSERT EDGE` statement inserts a (directed) edge from a source vertex (given by src_vid) to a destination vertex (given by dst_vid).
```
Comment (oldLady344) [inline docs-2.0/3.ngql-guide/11.edge-statements/1.insert-edge.md:15]: (directed)  why parentheses are used?

File: docs-2.0/3.ngql-guide/11.edge-statements/2.update-edge.md [hunk 23]
```diff
@@ -0,0 +1,24 @@
...
+
+Nebula Graph supports compare-and-set (CAS).
+
+> **NOTE:** `WHEN` and `YIELD` are optional.
+
+- `update_properties` is the properties of the edge to be updated. For example, `e1.col1 = $^.e1.col2 + 1` means to update `e1.col1` to `e1.col2+1`.
+
+    > **NOTE:**  `$^` indicates the edge to be updated.
+
+- `condition` is some constraints. Only when the `condition` met, `UPDATE` is executed successfully. `condition` supports expression operations.
```
Comment (oldLady344) [inline docs-2.0/3.ngql-guide/11.edge-statements/2.update-edge.md:17]: Only when the `condition` is met,

File: docs-2.0/3.ngql-guide/11.edge-statements/3.upsert-edge.md [hunk 24]
```diff
@@ -0,0 +1,42 @@
...
+```ngql
+UPSERT EDGE <src_vid> -> <dst_vid> [@rank] OF <edge_type> SET <update_properties> [WHEN <condition>] [YIELD <properties>]
+```
+
+- `update_properties` is the properties of the edge to be updated. For example, `e1.col1 = $^.e1.col2 + 1` means to update `e1.col1` to `e1.col2+1`.
+
+    > **NOTE:**  `$^` indicates the edge to be updated.
+
+- `condition` is some constraints. Only when the `condition` is met, `UPSERT` is executed successfully. `condition` supports expression operations.
+- `properties` specify the properties to be returned, `YIELD` returns the latest updated values.
```
Comment (oldLady344) [inline docs-2.0/3.ngql-guide/11.edge-statements/3.upsert-edge.md:12]: specifies. 

through the article, we treat all the parameters as a whole (singular noun), so keeping them consistent is recommended.

File: docs-2.0/3.ngql-guide/11.edge-statements/3.upsert-edge.md [hunk 25]
```diff
@@ -0,0 +1,42 @@
...
+
+`UPSERT` is a combination of `UPDATE` and `INSERT`. Use `UPSERT EDGE` to update properties on an edge if it exists or insert a new edge if it does not exist. The `UPDATE EDGE` statement only updates one edge at a time.
```
Comment (oldLady344) [inline docs-2.0/3.ngql-guide/11.edge-statements/3.upsert-edge.md:14]: only updates one edge at a time. --> updates only one edge at a time.

File: docs-2.0/3.ngql-guide/11.edge-statements/4.delete-edge.md (hunk 21 above)
Comment (oldLady344) [reply docs-2.0/3.ngql-guide/11.edge-statements/4.delete-edge.md:15]: > the edge whose edge type is serve, source vertex is "player100", destination vertex is "team200". --> the `serve` edge from "player100" to "team200"?

rewriting is recommended. To keep the sentence grammatically correct, you also can modify your version like this: This query deletes the edge of which the edge type is `serve`, the source vertex is "player100", and the destination vertex is "team200".

File: docs-2.0/3.ngql-guide/11.edge-statements/4.delete-edge.md [hunk 26]
```diff
@@ -0,0 +1,24 @@
...
+## Examples
+
+```ngql
+nebula> DELETE EDGE serve "player100" -> "team200";
+```
+
+This query deletes an serve edge from "player100" to "team200".
+
+```ngql
+nebula> GO FROM "player100" OVER follow YIELD follow._src AS src, follow._dst AS dst, follow._rank AS rank | \
```
Comment (whitewum) [inline docs-2.0/3.ngql-guide/11.edge-statements/4.delete-edge.md:18]: how about 
```
GO FROM "player100" OVER follow WHERE follow._dst == "team200" YIELD follow._src AS src, follow._dst AS dst, follow._rank AS rank | \
DELETE EDGE follow $-.src->$-.dst @ $-.rank;
```

Comment (whitewum) [reply docs-2.0/3.ngql-guide/11.edge-statements/4.delete-edge.md:18]: it is the same as delete all.

File: docs-2.0/3.ngql-guide/11.edge-statements/4.delete-edge.md [hunk 27]
```diff
@@ -0,0 +1,24 @@
...
+```ngql
+DELETE EDGE <edge_type> <src_vid> -> <dst_vid>[@<rank>] [, <edge_type> <src_vid> -> <dst_vid>[@<rank>] ...]
+```
+
+Use `DELETE EDGE` to delete edges. The `DELETE EDGE` statement deletes one edge or multiple edges at a time. You can use `DELETE EDGE` together with pipe. For more information about pipe, see [Pipe operator](../5.operators/4.pipe.md).
+
+## Examples
+
+```ngql
+nebula> DELETE EDGE serve "player100" -> "team200";
```
Comment (whitewum) [inline docs-2.0/3.ngql-guide/11.edge-statements/4.delete-edge.md:12]: it is the same as DELETE EDGE serve "player100" -> "team200" @ 0.

File: docs-2.0/3.ngql-guide/11.edge-statements/4.delete-edge.md (hunk 26 above)
Comment (Amber1990Zhang) [reply docs-2.0/3.ngql-guide/11.edge-statements/4.delete-edge.md:18]: I want to delete all here

File: docs-2.0/3.ngql-guide/11.edge-statements/4.delete-edge.md (hunk 27 above)
Comment (Amber1990Zhang) [reply docs-2.0/3.ngql-guide/11.edge-statements/4.delete-edge.md:12]: rank is not supposed to be documented here

**Review by whitewum — APPROVED** [review]

**Review by randomJoe211 — APPROVED** [review]

**Pull request merged**